src/v/rpc/rpc_compiler.py

```diff
@@ -224,8 +228,6 @@ class {{service_name}}_service_base<Codec>::failure_probes final : public finjec
     fast_prng _prng;
 };
 
-using {{service_name}}_service = {{service_name}}_service_base<rpc::default_message_codec>;
```

Comment:
@dotnwat - could be, but I do have a WIP which changes this specific generator to have a .h/.c split instead of doing it all in the .h, not sure if that would be harder with metaprogramming. I should get that one checked in...

Overall, I am not really against code-generation when it makes sense. Especially if you have _any_ code generation already, it it has already "infected" your build system so then the resistance to adding more is less.

Here are the advantages I see for code generation and metaprogramming respectively:

**Wins for code-generation:**

 - The actual generated code is usually much more understandable compared to a metaprogramming approach, or at least this is true if significant metaprogramming is involved.
 - I find writing python to do code generation easier than C++ meta-programming.
 - You can do "whatever you want": literally generate any C++ you can dream up, while metaprogramming is more limited, though obviously the scope has expanded a lot with new features.
 - Code generation is usually faster for C++ code, at least if you consider generating one file + compiling it, because code generation tends to be very fast, while compilation is slow and affected heavily by metaprogramming.

**Wins for doing it all in C++:**

- Simpler build: no dependency on the codegen language + runtime, no need for an extra step in the build, getting the dependencies correct, etc.
- Although codegen is usually faster "in isolation", it might slow things down because it is not as parallel (Amdahl's law) or not as incremental. In general things you do to improve the C++ build, like ccache or more fancy solutions, don't always play nice with codegen or at least don't help you.
- Using projects with codegen is more annoying as you need to run a build step to get all the files, e.g., if you need to work on an escalation and clone "random version" of the product, you can't start browsing without errors until you do the codegen step, since those headers are missing. Similarly it's a frequent source of errors when switch around between branches: for a .h that's checked it you always get the right version after a checkout but generated .h are dependent on codegen. Same comment applies to .c but .h are worse since they affect the compile for every file that includes them.
- Source control and review is probably easier with codegen: you can see the history of changes to a file, (i.e., "blame" works as you expect) - you can't with codegen because the generated files aren't checked in (though checking them in is a viable approach too), so you have to work back to the codegen and look for the blame there.

What else am I missing?

Reply:
I appreciate this write-up @travisdowns

> I find writing python to do code generation easier than C++ meta-programming.
> You can do "whatever you want"

These really resonate with me. The kafka protocol code generator has grown over time, but it was a major boon for us at the time, and still is.

> Wins for doing it all in C++:
> What else am I missing?

You're missing the most important one: being nerd sniped by constexpr/consteval. did you see that ray tracer that someone wrote with constexpr? haha.

I'm mostly joking around. For the RPC code generator my mental model is that the situation is quite simple and the generator isn't doing a whole lot. But that's not a big reason to replace it. I'm sure we'll revisit the rpc sub-system in general at some point and we could think about it.